mesonbuild/dependencies/base.py

```diff
@@ -501,51 +501,61 @@ def __init__(self, name, environment, kwargs, language=None):
+                # We never fallback if the user-specified option is no good, so
+                # stop returning options.
+                return
+            mlog.debug('Pkg-config binary missing from cross or native file, or env var undefined.')
             # Fallback on hard-coded defaults.
+            # TODO prefix this for the cross case instead of ignoring thing.
             if environment.machines.matches_build_machine(for_machine):
                 for potential_pkgpath in environment.default_pkgconfig:
-                    potential_pkgbin = self.check_pkgconfig(potential_pkgpath)
-                    if potential_pkgbin is None:
-                        mlog.debug(
-                            'default Pkg-config fallback %s for machine %s specified at %s but not found.',
-                            potential_pkgbin.name, for_machine, potential_pkgbin.command)
-                    else:
-                        return potential_pkgbin
-
-        self.pkgbin = search()
-        if self.pkgbin is None:
-            msg = 'Pkg-config binary for machine %s not found.' % for_machine
+                    mlog.debug('Trying a default pkg-config fallback at', potential_pkgpath)
+                    yield ExternalProgram(potential_pkgpath, silent=True)
+
+        # Only search for pkg-config for each machine the first time and store
+        # the result in the class definition
+        if PkgConfigDependency.class_pkgbin[for_machine] is False:
+            mlog.debug('Pkg-config binary for %s is cached missing.' % for_machine)
+        elif PkgConfigDependency.class_pkgbin[for_machine] is not None:
+            mlog.debug('Pkg-config binary for %s is cached.' % for_machine)
+        else:
+            assert PkgConfigDependency.class_pkgbin[for_machine] is None
+            mlog.debug('Pkg-config binary for %s is not cached.' % for_machine)
+            for potential_pkgbin in search():
+                mlog.debug(
+                    'Trying pkg-config binary %s for machine %s at %s.',
+                    potential_pkgbin.name, for_machine, potential_pkgbin.command)
+                version_if_ok = self.check_pkgconfig(potential_pkgbin)
+                if not version_if_ok:
+                    continue
+                if not self.silent:
+                    mlog.log('Found pkg-config:', mlog.bold(potential_pkgbin.get_path()),
+                             '(%s)' % version_if_ok)
+                PkgConfigDependency.class_pkgbin[for_machine] = potential_pkgbin
+                break
+            else:
+                if not self.silent:
+                    mlog.log('Found Pkg-config:', mlog.red('NO'))
+                # Set to False instead of None to signify that we've already
+                # searched for it and not found it
+                PkgConfigDependency.class_pkgbin[for_machine] = False
+
+        self.pkgbin = PkgConfigDependency.class_pkgbin[for_machine]
+        if self.pkgbin == False:
```

Comment:
__[Flake8]__ [E712] comparison to False should be 'if cond is False:' or 'if not cond:'

<sub>posted by Sider</sub>